actionpack/lib/action_dispatch/routing/route_set.rb

```diff
@@ -693,20 +719,43 @@ def generate(route_key, options, recall = {})
                           :original_script_name]
 
       def optimize_routes_generation?
-        default_url_options.empty?
+        !default_url_options
       end
 
       def find_script_name(options)
-        options.delete(:script_name) || ''
```

Comment:
We should have. I merged this code today with tests https://github.com/rails/rails/commit/44ff0313c121f528a68b3bd21d6c7a96f313e3d3